### PR TITLE
support numeric literals with underscores

### LINKED ISSUE #28967: Should we support numeric literals in form of 1_000_000? (with underscore separator of groups).
Pros:
- it is convenient to type and read numbers this way;
- Ada, Perl, C++, Go and Rust already support that;
- can support thousands as well as lakhs and crores;

Cons:
- standard SQL does not have this support;
- PostgreSQL treat `SELECT 1_000_000` as number one with `_000_000` alias: http://sqlfiddle.com/#!17/0a28f/400
- C++ has `1'000'000` instead of `1_000_000`;

What do you think?

### PR DESCRIPTION
**Pull request by unbyte**
### Changelog category:
- Improvement

### Changelog entry:
Support numeric literals with underscores. closes #28967

### CODE REVIEW

**Comment by qoega** [comment]
No hurry at all. You can even wait for existing perf test results in CI first. We just don't want to have performance degradation for previously supported literals.

**Review by vdimir — COMMENTED** [review]
Thank you for your changes!

I'd suggest add stateless tests, simple queries like `SELECT 1_000;` `SELECT number < 10_000 from numbers(1_000_000);`  something with `EXPLAIN SYNTAX` and so on.

And also check some corner cases, like `_1000`, `1____0`, `10_`, it shouldn't be parsed. Lets keep the same semantics that implemented in python3 for example.

**Comment by alexey-milovidov** [comment]
I think leading, trailing, and double underscores should not be supported.

**Comment by vdimir** [comment]
> Python has a more strict rule, should we take it as the standard or use a looser rule?

@unbyte Yes, I'd also vote for the strict rules, as Alexey said.

So `_1000`, `1____0`, `10_`, `+_1`, `-_1`, `1_e5`, `1e_5`, `0_x2` should not be parsed

**Comment by alexey-milovidov** [comment]
Anything beginning with a number should not be parsed as a bareword.

AFAIK it was added a long time ago for MaterializedMySQL and no one noticed (but it should not be added). We can remove it even if it will break some corner cases with MaterializedMySQL.

**Comment by vdimir** [comment]
@alexey-milovidov maybe we need compatibility flag for that, what do you think?

**Comment by tavplubix** [comment]
> Anything beginning with a number should not be parsed as a bareword.
> 
> AFAIK it was added a long time ago for MaterializedMySQL and no one noticed (but it should not be added). We can remove it even if it will break some corner cases with MaterializedMySQL.

Yes, it was added in #14232 (that fixed #14231) and maybe it was merged by mistake. It works a bit weird way, for example, `42e` and `0x` are barewords, but `42e1` and `0xd` are numeric literals. I think it's ok to drop support for this. But maybe we can add some flag to Lexer that enables paring of identifiers starting from numbers and pass this flag from MaterializedMySQL.

File: tests/queries/0_stateless/02354_numeric_literals_with_underscores.sql [hunk 1]
```diff
@@ -0,0 +1,6 @@
+select 1_000;
+select 1.00_00_01;
+select 1.000_0001e2;
+select 0x12_34_56_78;
+select 0x12_34_56_78p1;
+select 0b0010_0100_0111;
```
Comment (vdimir) [inline tests/queries/0_stateless/02354_numeric_literals_with_underscores.sql:6]: Let's also add negative testcases
Maybe you'll find more corner cases

```suggestion
select 0b0010_0100_0111;

select _1000; -- { serverError UNKNOWN_IDENTIFIER }
select _1000 FROM (SELECT 1 AS _1000) FORMAT Null;

select -_1; -- { serverError UNKNOWN_IDENTIFIER }
select -_1 FROM (SELECT -1 AS _1) FORMAT Null;

select +_1; -- { clientError SYNTAX_ERROR }

select 1____0; -- { clientError SYNTAX_ERROR }
select 10_; -- { clientError SYNTAX_ERROR }
select 1_e5; -- { clientError SYNTAX_ERROR }
select 1e_5; -- { clientError SYNTAX_ERROR }
select 1e_; -- { clientError SYNTAX_ERROR }
select 0_x2; -- { clientError SYNTAX_ERROR }
select 0x_; -- { clientError SYNTAX_ERROR }

```

**Review by vdimir — COMMENTED** [review]
I've tested new changes a bit, and found following: 

For some reason, `select 0x` results to `SYNTAX_ERROR` but `select 0x;` with `;` at the end doesn't

```
$ clickhouse client -q "select 0x"
Code: 62. DB::Exception: Syntax error: failed at position 8 ('0x'): 0x. Wrong number: '0x'. (SYNTAX_ERROR)

$ clickhouse client -q "select 0x;"
0

$ ~/clickhouse client -q "select 42e"
Code: 62. DB::Exception: Syntax error: failed at position 8 ('42e'): 42e. Wrong number: '42e'. (SYNTAX_ERROR)

$ ~/clickhouse client -q "select 42e;"
42

```

Also, behaviour is inconsistent between `+` and `-` followed by literal starting from underscore.
```
select -_1; -- { serverError UNKNOWN_IDENTIFIER }
select +_1; -- { clientError SYNTAX_ERROR }
```

**Comment by vdimir** [comment]
@unbyte Yeah, it's fine. No worries at all 😉

**Review by vdimir — APPROVED** [review]
I going to merge this PR as well to preserve commits and add @unbyte  to the list of contributors

**Pull request merged**